clash-protocols/src/Protocols/BiDf.hs

```diff
@@ -0,0 +1,141 @@
+{-# OPTIONS_GHC -fplugin Protocols.Plugin #-}
+
+-- | Bi-directional request/response-style 'Df' channels.
+module Protocols.BiDf (
+  BiDf,
+    -- * Conversion
+  fromDfs,
+  toDfs,
+  fromBiDf,
+  toBiDf,
+    -- * Trivial combinators
+  void,
+  loopback,
+    -- * Mapping
+  dimap,
+  -- * Fan-in
+  fanin
+) where
+
+import Prelude ()
+
+import Clash.Prelude
+
+import Protocols
+import qualified Protocols.Df as Df
+
+-- | A 'Protocol' allowing requests to be passed downstream, with corresponding
+-- responses being passed back upstream. Responses are provided in the order that
+-- their corresponding requests were submitted.
+--
+-- *Correctness conditions*
+--
+--  - The response channel must not produce a value before the request channel
+--    has produced a value. The response may be produced in the same cycle the
+--    request is acknowledged (but see the law about a combinational path
+--    below).
+--
+--  - Each request must be paired with exactly one response.
+--
+--  - Responses must be issued in the order that their corresponding requests arrived.
+--
+--  - Both the request and response channels must obey usual 'Df' correctness
+--    conditions.
+--
+--  - There must not be a combinational path from the request channel to the
+--    response channel.
```

Comment:
Am I reading correctly that this disallows stateless circuits such as:
```hs
idBiDf :: Circuit (BiDf dom a a) ()
idBiDf = Circuit go
 where
  go ((fwd, bwd), _) = ((bwd, fwd), ())
```

If so, why would that not be allowed?